src/main/java/com/limechain/runtime/Runtime.java

```diff
@@ -63,87 +72,40 @@ private byte[] callInner(String functionName, RuntimePointerSize parameterPtrSiz
-     * and set it afterward via this setter.
-     */
-    void setVersion(RuntimeVersion runtimeVersion) {
-        this.version = runtimeVersion;
-    }
-
-    public int getHeapBase() {
-        return instance.exports.getGlobal(WasmExports.HEAP_BASE.getValue()).getIntValue();
-    }
-
-    public int getDataEnd() {
-        return instance.exports.getGlobal(WasmExports.DATA_END.getValue()).getIntValue();
-    }
-
-    private Memory getMemory() {
-        return instance.exports.getMemory(WasmExports.MEMORY.getValue());
+        return context.getSharedMemory().readData(responsePtrSize);
     }
 
     /**
-     * Get the data stored in memory using a {@link  RuntimePointerSize}
-     *
-     * @param runtimePointerSize pointer to data and its size
-     * @return byte array with read data
+     * @return the {@link RuntimeVersion} of this runtime instance.
+     * @implNote The runtime version is cached in the context, so no actual runtime invocation is performed.
      */
-    public byte[] getDataFromMemory(RuntimePointerSize runtimePointerSize) {
-        ByteBuffer memoryBuffer = this.getMemory().buffer();
-        byte[] data = new byte[runtimePointerSize.size()];
-        memoryBuffer.position(runtimePointerSize.pointer());
-        memoryBuffer.get(data);
-        return data;
+    public RuntimeVersion getVersion() {
+        return this.context.getRuntimeVersion();
     }
 
-    /**
-     * Write data to memory, by allocating space in memory and then writing to it.
-     *
-     * @param data data to be written
-     * @return a pointer size to the written data
-     */
-    public RuntimePointerSize writeDataToMemory(byte[] data) {
-        RuntimePointerSize allocatedPointer = allocate(data.length);
-        writeDataToMemory(data, allocatedPointer);
-        return allocatedPointer;
+    RuntimeVersion callCoreVersion() {
+        return ScaleUtils.Decode.decode(this.call("Core_version"), new RuntimeVersionReader());
     }
 
     /**
-     * Write data to memory, by using a {@link RuntimePointerSize}.
-     * <br>Data will be written at the given {@link RuntimePointerSize#pointer() pointer}.
-     * <br>Only the first bytes up to the given {@link RuntimePointerSize#size() size} will be written.
-     *
-     * @param data               data to be written to memory
-     * @param runtimePointerSize pointer to memory and size of data to be stored.
+     * Closes the underlying instance.
      */
-    public void writeDataToMemory(byte[] data, RuntimePointerSize runtimePointerSize) {
-        ByteBuffer memoryBuffer = getMemory().buffer();
-        memoryBuffer.position(runtimePointerSize.pointer());
-        memoryBuffer.put(data, 0, runtimePointerSize.size());
+    public void close() {
+        // TODO:
+        //  We're not sure whether we should also close the module.
+        //  That's what Gossamer does though, but their wasm runtime could be doing things differently underneath.
+        this.module.close();
```

Comment:
Gossamer are irrelevant here. The method deletes the module pointer which is created because it is created in the rust part of wasmer-java. Have to manually test impact.

Reply:
Virtually the same topic as https://github.com/LimeChain/Fruzhin/pull/456#discussion_r1622192678